### PR TITLE
fix(dashboard): de-duplicate rows via row_signature, graceful shutdown, collapsible panel (#605)

### PR DESCRIPTION
**Pull request by danielgerlag**
## Summary

Fixes duplicate rows in the dashboard reaction (#605) by keying rows on the engine's
canonical identity, plus two related dashboard improvements.

### 1. Duplicate rows (#605) — keyed by `row_signature` end-to-end
The canonical row identity in drasi is `row_signature` (engine-stamped on every
`ResultDiff`; the query result state, live-results store, and outbox all key by it).
The dashboard discarded it and reinvented a weaker `id`/JSON-equality key, and the
snapshot/bootstrap path dropped the signature entirely — **including across FFI**, which
is why #605 reproduced on drasi-server (the plugin path). With the initial snapshot and
the live diff stream in different identity spaces, the browser couldn't dedupe → duplicates.

- **lib**: `SnapshotResponse::stream_keyed` + keyed `SnapshotStream` APIs
  (`from_keyed_stream`/`collect_keyed_vec`/`next_keyed`); the public `Stream` impl still
  yields bare values so existing consumers are untouched. Shared `KeyedSnapshotRow { k, v }`
  wire envelope.
- **FFI (host + plugin)**: snapshot rows cross FFI as `KeyedSnapshotRow` so the signature
  survives; the plugin parses it back (falling back to signature `0` for bare payloads).
  **`FFI_SDK_VERSION` bumped `0.9.1 → 0.10.0`** — the vtable layout is unchanged, but the
  per-row wire format changed semantically, so mismatched host/plugin must be rebuilt.
- **dashboard server**: `WebSocketResultDiff` carries `row_signature` (`k`); snapshot rows
  are `{ row_signature, data }`; `apply`/`seed_rows`/`find_row_index` key by signature,
  falling back to the `id`/equality heuristic only when the signature is unknown (`0`).
- **browser**: keys rows by `row_signature` (stored as a non-enumerable `__sig` so it never
  leaks into rendering); the snapshot loader threads the signature through; heuristic retained
  as fallback.

> Note: this supersedes the earlier upsert-by-`id` band-aid that was the first commit on this
> branch. The `id`/equality logic is kept only as a fallback for signature-less rows.

### 2. Stale browser tab after reaction restart
`axum::serve` ran with no graceful shutdown and `stop()` only aborted the accept loop, so
existing keep-alive/WebSocket connections stayed alive in-process and a refreshed tab reused
a pooled connection to the old server. Now driven by a `watch` shutdown signal
(`with_graceful_shutdown`), WebSocket handlers close on shutdown, and assets are served with
`Cache-Control: no-cache`.

### 3. Collapsible "Available Queries" panel
The editor's left sidebar can be collapsed/expanded; state persisted in `localStorage`, charts
reflow on toggle.

## Testing
- `cargo test -p drasi-reaction-dashboard` — 57 unit + 2 integration tests (incl. graceful-shutdown
  regression and the real-query WS round-trip exercising `row_signature`).
- `cargo test -p drasi-lib` snapshot-stream signature-preservation tests; `host-sdk`/`plugin-sdk`
  lib unit tests (incl. the FFI snapshot bridge).
- JS: `tests/apply_result_diff.test.js` (signature keying + id/equality fallback) and existing
  handlebars tests, wired into `make test-js`.
- `cargo build --workspace`; clippy `-D warnings` and `cargo fmt --check` clean on touched crates.

> Pre-existing `host-sdk` `integration_test.rs` cases that require prebuilt cdylib artifacts
> ("SKIP: … not built as cdylib") fail identically on a clean tree and are unrelated.

### CODE REVIEW

**Comment by danielgerlag** [comment]
Thanks for the review — all three comments addressed in c558e8e2:

1. **JS test VM loader (`export { ... }`)** — the strip step now removes `export { ... };` list statements entirely before the generic `^export` rule, so re-exports (e.g. `export { AGGREGATION_MODES };`) can't leave invalid script. Applied to both `apply_result_diff.test.js` and `handlebars_helpers.test.js`.
2. **Hard-coded port 19111** — the shutdown integration test now reserves an ephemeral port (`bind 127.0.0.1:0`, read `local_addr().port()`, drop) and builds all URLs from it, so it's collision-resistant under parallel runs.
3. **`serve_asset` per-request allocations** — `Content-Type`/`Cache-Control` are now passed as `&str` instead of `to_string()`.

Note: the PR has since grown — the duplicate-row fix was reworked to key rows by the engine's canonical `row_signature` end-to-end (incl. the FFI snapshot path, with an `FFI_SDK_VERSION` bump), superseding the earlier upsert-by-`id` approach the review summary describes.

File: components/reactions/dashboard/tests/apply_result_diff.test.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  RegExp,
+  JSON,
+  console,
+};
+
+// Strip ES module export/import syntax so it can run as a script in vm.
+const strippedSource = widgetsSource
+  .replace(/^export\s+/gm, "")
+  .replace(/^import\s+.*$/gm, "");
+
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/reactions/dashboard/tests/apply_result_diff.test.js:36]: Fixed in c558e8e2: the strip step now removes `export { ... };` list statements entirely (e.g. `export { AGGREGATION_MODES };`) before the generic `^export` rule, so re-exports can't leave invalid script. Applied here and in `handlebars_helpers.test.js`.

File: components/reactions/dashboard/tests/integration_tests.rs [hunk 2]
```diff
@@ -233,3 +233,120 @@ async fn test_dashboard_reaction_end_to_end() -> Result<()> {
...
+    let query = Query::cypher("shutdown-query")
+        .query("MATCH (p:Person) RETURN p.name AS name")
+        .from_source("shutdown-source")
+        .auto_start(true)
+        .build();
+
+    let reaction = DashboardReaction::builder("shutdown-dashboard")
+        .with_query("shutdown-query")
+        .with_port(19111)
+        .build()?;
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/reactions/dashboard/tests/integration_tests.rs:271]: Fixed in c558e8e2: the test now reserves an ephemeral port (`bind 127.0.0.1:0`, read `local_addr().port()`, drop the listener) via a `reserve_port()` helper and builds all URLs from it, so it's collision-resistant under parallel runs.

File: components/reactions/dashboard/src/dashboard.rs [hunk 3]
```diff
@@ -106,7 +113,19 @@ fn serve_asset(path: &str) -> Response {
...
-            ([(header::CONTENT_TYPE, mime)], content.data.into_owned()).into_response()
+            // Force revalidation so a browser tab can't serve a stale shell/JS
+            // from its in-memory cache after the reaction is restarted.
+            (
+                [
+                    (header::CONTENT_TYPE, mime.to_string()),
+                    (
+                        header::CACHE_CONTROL,
+                        "no-cache, no-store, must-revalidate".to_string(),
+                    ),
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/reactions/dashboard/src/dashboard.rs]: Fixed in c558e8e2: `Content-Type` and `Cache-Control` are now passed as `&str` (the cache value is a `&'static str`), removing the per-request `String` allocations.

File: components/reactions/dashboard/src/websocket.rs [hunk 4]
```diff
@@ -269,12 +353,35 @@ impl QuerySnapshotStore {
...
+}
+
+/// Find a row by its canonical identity.
+///
+/// Matches on `row_signature` (the engine-stamped identity) when it is known
+/// (non-zero); otherwise falls back to matching by the `id` field, then by full
+/// data equality.
+fn find_row_index(rows: &[SnapshotRow], row_signature: u64, data: &Value) -> Option<usize> {
+    if row_signature != 0 {
+        return rows.iter().position(|r| r.row_signature == row_signature);
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/reactions/dashboard/src/websocket.rs]: Fixed in bda894b6: `find_row_index` now falls through to the id/equality search when the signature match misses, so a row seeded with signature 0 is replaced (not duplicated) by a later real-signature diff. Mirrored on the client in `findRowIndexForDiff`, and added Rust (`test_snapshot_store_sig0_seed_then_real_signature_upserts`) + JS regression tests. (In practice the version gate blocks an old sig-0 plugin from loading into this host, but the fallthrough makes it correct regardless.)

File: components/reactions/dashboard/static/js/dashboard.js [hunk 5]
```diff
@@ -534,8 +534,12 @@ export class DashboardDesigner {
         const runtime = this.runtimeByWidgetId.get(widget.id) ?? createWidgetRuntime();
         // Only populate if the runtime has no data yet
         if (runtime.rows.length === 0 && runtime.aggregation === null) {
-          for (const row of rows) {
-            applyResultDiff(runtime, { op: "add", data: row });
+          for (const entry of rows) {
+            // Snapshot rows arrive as { k: row_signature, v: data }; pass the
+            // signature through so live diffs match by identity (issue #605).
+            const data = entry && typeof entry === "object" && "v" in entry ? entry.v : entry;
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/reactions/dashboard/static/js/dashboard.js]: Fixed in bda894b6: the loader now treats an entry as the `{k, v}` envelope only when **both** `k` and `v` are present, so a plain row containing a `v` field isn't misread.

File: components/reactions/dashboard/static/js/widgets.js [hunk 6]
```diff
@@ -343,25 +375,37 @@ function normalizeDiffData(diff) {
 export function applyResultDiff(runtime, diff) {
   if (!runtime) return;
 
+  const sig = rowSig(diff.k ?? diff.row_signature);
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/reactions/dashboard/static/js/widgets.js]: Fixed in bda894b6: dropped the dead branch — `const sig = rowSig(diff.k);`.

File: components/reactions/dashboard/src/websocket.rs [hunk 7]
```diff
@@ -37,6 +37,10 @@ pub enum ClientMessage {
 #[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
 pub struct WebSocketResultDiff {
     pub op: String,
+    /// The row's engine-stamped `row_signature` — the canonical row identity.
+    /// Serialized as `k`; `0` means unknown (clients fall back to id/equality).
+    #[serde(rename = "k", default)]
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/reactions/dashboard/src/websocket.rs]: Fixed in 6c442610: `row_signature` is now serialized as a JSON **string** (`k`) on both the WS diff and the snapshot payload, so the full 64 bits survive in the browser; the client keys by the string (no `BigInt`/arithmetic needed since we only compare for identity). Deserialization also accepts a numeric form for resilience. Added a round-trip test with a value > 2^53 and a JS test that two near-2^53 signatures aren't conflated.

File: components/reactions/dashboard/src/dashboard.rs [hunk 8]
```diff
@@ -283,6 +299,33 @@ impl Reaction for DashboardReaction {
...
+        // Seed the accumulated snapshot store from the bootstrap snapshot so the
+        // live WebSocket stream starts from the same baseline the fallback fetcher
+        // reads, instead of starting empty (see issue #605). Seeding is best-effort:
+        // failures (e.g. bootstrap not yet complete) are logged and skipped, and the
+        // fallback fetcher in the API still serves initial state.
+        if let Some(fetcher) = snapshot_fetcher.as_ref() {
+            for query_id in &self.base.queries {
+                match fetcher.fetch_snapshot(query_id).await {
+                    Ok(stream) => {
+                        let rows = stream.collect_keyed_vec().await;
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/reactions/dashboard/src/dashboard.rs]: Fixed in 6c442610: bootstrap seeding now collects at most `max_rows_per_query` rows via the new `SnapshotStream::collect_keyed_vec_capped(limit)`, bounding peak memory. The per-row `find_row_index` scan remains, but N is now capped by the limit, so the one-time seeding cost is bounded.

File: lib/src/queries/output_state.rs [hunk 9]
```diff
@@ -24,10 +24,23 @@ use std::collections::VecDeque;
...
 use tokio_stream::Stream;
 
 use crate::channels::{QueryResult, ResultDiff};
 
+/// Wire envelope pairing a row's engine-stamped `row_signature` with its data.
+///
+/// Used as the per-row serialization format when snapshot rows cross the FFI
+/// boundary, so the canonical row identity survives instead of being dropped.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct KeyedSnapshotRow {
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply lib/src/queries/output_state.rs:37]: Declined (with rationale): `ffi-primitives` is intentionally dependency-free — its docs state "No domain-specific dependencies — this crate depends only on std" — and `KeyedSnapshotRow` needs `serde`/`serde_json::Value`. `drasi-lib` already depends on serde heavily (e.g. `QueryResult`/`ResultDiff` derive Serialize/Deserialize), so it adds no new dependency there, and both SDKs already depend on `drasi-lib`. Keeping it in lib avoids pulling serde into the std-only primitives crate.

File: components/reactions/dashboard/src/websocket.rs [hunk 10]
```diff
@@ -178,10 +197,22 @@ impl WebSocketHub {
     }
 }
 
+/// A single accumulated snapshot row paired with its canonical identity.
+///
+/// `row_signature` is the engine-stamped identity (`0` when unknown, in which
+/// case matching falls back to an `id`/equality heuristic on `data`).
+#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
+pub struct SnapshotRow {
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/reactions/dashboard/src/websocket.rs:238]: Kept separate (with rationale): after the security fix, the dashboard's `SnapshotRow` serializes `k` as a **string** (JS 64-bit precision-safety for the browser), whereas `KeyedSnapshotRow` is a Rust↔Rust FFI envelope that keeps `k` as a `u64` number. Same shape, different serialization contracts — merging them would force one side's wire format onto the other.

File: lib/src/queries/output_state.rs [hunk 11]
```diff
@@ -365,7 +416,10 @@ impl Stream for SnapshotStream {
         mut self: Pin<&mut Self>,
         cx: &mut std::task::Context<'_>,
     ) -> std::task::Poll<Option<Self::Item>> {
-        self.inner.as_mut().poll_next(cx)
+        self.inner
+            .as_mut()
+            .poll_next(cx)
+            .map(|opt| opt.map(|(_, v)| v))
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply lib/src/queries/output_state.rs:445]: Addressed in 6c442610 with documentation: the `Stream` impl is retained because existing value-only consumers (e.g. sse / snapshot-test) iterate it via `StreamExt`, but it now carries a doc comment stating it drops signatures and that identity-preserving consumers must use `next_keyed`/`collect_keyed_vec` (and noting this lossy path is what caused #605). Removing the impl outright would be a larger breaking change across those crates.

File: components/reactions/dashboard/src/dashboard.rs [hunk 12]
```diff
@@ -53,6 +53,11 @@ pub struct DashboardReaction {
     websocket_hub: WebSocketHub,
     snapshot_store: QuerySnapshotStore,
     task_handles: Arc<tokio::sync::Mutex<Vec<JoinHandle<()>>>>,
+    server_handle: Arc<tokio::sync::Mutex<Option<JoinHandle<()>>>>,
+    /// Signals the HTTP/WebSocket server to shut down gracefully. `true` means
+    /// "shut down now": the axum server stops accepting and closes idle
+    /// keep-alive connections, and WebSocket handlers terminate.
+    shutdown_tx: tokio::sync::watch::Sender<bool>,
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/reactions/dashboard/src/dashboard.rs:61]: Acknowledged (nit) — kept as-is. `stop()` is the single coordinated entry point and signals both the base channel (processing/heartbeat) and the server's `watch`; per-connection WebSocket handlers each need their own receiver, which the `watch` provides cleanly (the base channel is consumed once by the processing task). Fully unifying them would require changing the base shutdown-channel semantics, which is out of scope here.

File: components/reactions/dashboard/src/dashboard.rs [hunk 13]
```diff
@@ -283,6 +299,33 @@ impl Reaction for DashboardReaction {
             .await
             .and_then(|ctx| ctx.snapshot_fetcher.clone());
 
+        // Seed the accumulated snapshot store from the bootstrap snapshot so the
+        // live WebSocket stream starts from the same baseline the fallback fetcher
+        // reads, instead of starting empty (see issue #605). Seeding is best-effort:
+        // failures (e.g. bootstrap not yet complete) are logged and skipped, and the
+        // fallback fetcher in the API still serves initial state.
+        if let Some(fetcher) = snapshot_fetcher.as_ref() {
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/reactions/dashboard/src/dashboard.rs]: Good catch — fixed in 6c442610: seeding now runs **before** the processing task is spawned, so a concurrently delivered `Delete` is applied after seeding (correctly removing the row) rather than racing it and being resurrected by the stale snapshot. Comment updated to explain the ordering.

File: components/plugin-sdk/src/ffi/vtable_gen.rs [hunk 14]
```diff
@@ -184,12 +184,21 @@ pub(crate) fn make_snapshot_stream(
...
-                        }
+                    // Rows cross FFI as a `KeyedSnapshotRow` envelope so the
+                    // canonical `row_signature` identity survives. Fall back to
+                    // treating the payload as a bare row (signature 0) for
+                    // forward/backward resilience.
+                    match serde_json::from_str::<drasi_lib::queries::output_state::KeyedSnapshotRow>(
+                        &json_str,
+                    ) {
+                        Ok(envelope) => Some((envelope.k, envelope.v)),
+                        Err(_) => match serde_json::from_str::<serde_json::Value>(&json_str) {
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/plugin-sdk/src/ffi/vtable_gen.rs:235]: Fixed in 6c442610: added `snapshot_stream_tests` in `vtable_gen.rs` — `make_snapshot_stream_falls_back_to_bare_row` feeds a bare `{"id":1}` (no envelope) and asserts exactly one item, signature 0, and the bare value; `make_snapshot_stream_parses_keyed_envelope` covers the normal path. (Also: the e2e test now asserts a non-zero string `k` on the live add diff, and uses an ephemeral port — the two items from the review summary.)

File: components/reactions/dashboard/static/js/dashboard.js [hunk 15]
```diff
@@ -534,8 +534,12 @@ export class DashboardDesigner {
...
+            const k = entry && typeof entry === "object" ? entry.k : undefined;
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/reactions/dashboard/static/js/dashboard.js]: Fixed in 6c442610: extracted the `{k, v}` decode into an exported pure helper `snapshotRowToAddDiff(entry)` in widgets.js (now used by `fetchSnapshot`), and added JS tests covering envelope decode, plain-row passthrough (a row containing a `v` field isn't misread), and the snapshot-load → live-diff signature threading.

File: lib/src/queries/output_state.rs [hunk 16]
```diff
@@ -332,17 +353,35 @@ impl SnapshotStream {
...
+    ) -> Self {
+        use tokio_stream::StreamExt;
+        Self {
+            inner: Box::pin(stream.map(|v| (0u64, v))),
+            as_of_sequence,
+            config_hash,
+        }
+    }
+
+    /// Create a `SnapshotStream` from a stream of `(row_signature, value)` pairs.
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply lib/src/queries/output_state.rs:380]: Fixed in 6c442610.

File: lib/src/queries/output_state.rs [hunk 17]
```diff
@@ -351,11 +390,23 @@ impl SnapshotStream {
         }
     }
 
-    /// Collect all rows from the stream into a `Vec`.
+    /// Collect all rows from the stream into a `Vec`, dropping signatures.
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply lib/src/queries/output_state.rs:397]: Fixed in 6c442610.

File: lib/src/queries/output_state.rs [hunk 18]
```diff
@@ -24,10 +24,23 @@ use std::collections::VecDeque;
...
+///
+/// Used as the per-row serialization format when snapshot rows cross the FFI
+/// boundary, so the canonical row identity survives instead of being dropped.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct KeyedSnapshotRow {
+    /// The row's `row_signature` (engine-computed identity). `0` means unknown.
+    pub k: u64,
+    /// The row data.
+    pub v: serde_json::Value,
+}
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply lib/src/queries/output_state.rs:42]: Fixed in 6c442610.

File: components/reactions/dashboard/src/dashboard.rs [hunk 19]
```diff
@@ -53,6 +53,11 @@ pub struct DashboardReaction {
     websocket_hub: WebSocketHub,
     snapshot_store: QuerySnapshotStore,
     task_handles: Arc<tokio::sync::Mutex<Vec<JoinHandle<()>>>>,
+    server_handle: Arc<tokio::sync::Mutex<Option<JoinHandle<()>>>>,
```
[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/reactions/dashboard/src/dashboard.rs:57]: Fixed in 6c442610.

**Comment by danielgerlag** [comment]
## ⚠️ FFI version coordination with #608

Heads-up for whoever merges second: **#607 and #608 both make independent, ABI-breaking FFI changes and both currently land the gate version on `0.10.0` — via different mechanisms.** Because the loader only compares `major.minor`, two different `0.10.0` ABIs would pass the gate against each other, which is exactly the silent mismatch the gate exists to prevent.

| | #607 (this PR) | #608 (#602) |
|---|---|---|
| FFI version mechanism | `FFI_SDK_VERSION = env!("CARGO_PKG_VERSION")`, driven by a **workspace version bump** `0.9.1 → 0.10.0` | **Decouples** it — deletes the `env!` form and hardcodes `pub const FFI_SDK_VERSION = "0.10.0"` |
| ABI change | snapshot rows cross the (pre-existing, JSON-string) snapshot iterator as a `{ "k": <row_signature>, "v": <row> }` envelope instead of a bare row | source-change / bootstrap / `QueryResult` events cross as MessagePack bytes instead of reinterpreted `repr(Rust)` opaque pointers |
| Net gate version | `0.10.0` | `0.10.0` |

### What the second PR to merge must do
1. **Bump the FFI minor again — to `0.11.0`** — so the *combined* ABI gets a version distinct from the first-landed `0.10.0`. (A `0.10.0` host carrying only one of the two changes must not accept a `0.10.0` plugin built with the other.)
2. **Converge on #608's mechanism.** #608 intentionally decouples `FFI_SDK_VERSION` from `CARGO_PKG_VERSION` and documents it in the FFI `agents.md`. If #608 lands first, this PR should **drop the `env!`/workspace-version approach and set the hardcoded constant to `0.11.0`** (plus a `History:` line for the snapshot wire-format change). The `0.9.1 → 0.10.0` workspace bump here then becomes cosmetic for the gate and can stay or be reverted.
3. **Expect textual merge conflicts** in the two files both PRs edit:
   - `components/plugin-sdk/src/ffi/vtable_gen.rs`
   - `components/host-sdk/src/proxies/reaction.rs`
   They touch *different* functions (here: `make_snapshot_stream` / `snapshot_iter_next`; there: event-payload serialization + the new `payload.rs`), so the changes are functionally independent — just adjacent.

### Functional interaction: none
The two changes are on **separate FFI paths**:
- This PR's snapshot rows already crossed as JSON strings via `FfiOwnedStr` (owned/copied bytes — no `repr(Rust)` pointer reinterpret); we only added the `k` field to that string. #608 does not touch the snapshot iterator.
- #608's MessagePack change is for the hot source-change/`QueryResult` event path (the actual #602 corruption vector).

No behavioral conflict — this is purely a version-gate-coordination + merge-conflict note.

cc: #608

**Review by agentofreality — APPROVED** [review]

**Review by ruokun-niu — APPROVED** [review]
